internal/rotate_est.go

```diff
@@ -0,0 +1,178 @@
+package internal
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	oidKeyUsage         = asn1.ObjectIdentifier([]int{2, 5, 29, 15})
+	oidExtendedKeyUsage = asn1.ObjectIdentifier([]int{2, 5, 29, 37})
+
+	asn1DigitalSignature = []byte{3, 2, 7, 128}
+	asn1TlsWebClientAuth = []byte{48, 10, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2}
+)
+
+type estStep struct{}
+
+func (s estStep) Name() string {
+	return "Generate new certificate"
+}
+
+func (s estStep) Completed(state CertRotationState) bool {
+	return len(state.NewKey) > 0 && len(state.NewCert) > 0
+}
+
+func (s estStep) Execute(handler *CertRotationHandler) error {
+	// Find the current certificate so we can build the proper EST request
+	tlsCert := handler.client.Transport.(*http.Transport).TLSClientConfig.Certificates[0]
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil
+	}
+
+	var signer crypto.Signer
+	var newKey string
+
+	// Generate a new private key
+	if handler.usePkcs11() {
+		newKey = "07"
+		signer, err = handler.crypto.ctx.GenerateECDSAKeyPair(idToBytes(newKey), elliptic.P256())
+		if err != nil {
+			return fmt.Errorf("Unable to generate new keypair in HSM: %w", err)
+		}
+	} else {
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return fmt.Errorf("Unable to generate new private key: %w", err)
+		}
+		keyBytes, err := x509.MarshalECPrivateKey(key)
+		if err != nil {
+			return fmt.Errorf("Unable to serialize new private key: %w", err)
+		}
+		keyBytes = pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: keyBytes})
+		newKey = string(keyBytes)
+		signer = key
```

Comment:
A bit confusing name `signer`, it doesn't sign anything.

Reply:
It's data type is crypto.Signer but it also signs the CSR when passed to `createB64CsrDer`

Reply:
Hm, I thought the CSR is signed by the `simplereenroll` handler.

Reply:
the handler signs the actual certificate. but when you create a CSR you include proof you posses the public key included in it - https://pkg.go.dev/crypto/x509#CreateCertificateRequest

Reply:
ok, I understand it, however it doesn't make the name `signer` clearer to me :).
Anyway, it's not important.